### ISSUE TITLE
Sonar originated altitude is (almost) always lower by few cm

### ISSUE THREAD

**Issue by jflyper** [iNavFlight/inav#621]
As I was testing the sonar support for OMNIBUS F3, I came to notice that the distance reported by sonar is always shorter than the actual distance by 1~3cm.

I took a quick look at the code, and found at least a couple of places where loss of significant digit is occurring (accounts for 2cm).

I actually don't know if sonar is requested to generate a distance in this precision.

Any thoughts?

Oh, and it came to me that we should have a variable to adjust the altitude based on the relative position of where the sonar is mounted on a vehicle.

**Comment by bk79** [comment]
Consider that sonar is reliable below 3m and baro is 0.25 m precise... 2 cm
is quite precise for me

**Comment by jflyper** [comment]
Okay. Then I would look into the occasional jump; 350cm at 10cm and a like.

**Comment by bk79** [comment]
Yes those glitchesvare something to look into personally i would use baro
to have a reliabilty threshold

**Comment by jflyper** [comment]
Shall we open a new issue?

**Comment by jflyper** [comment]
@bk79 @martinbudden 

What do you think of this? betaflight/betaflight#1212

**Comment by bk79** [comment]
The idea of having a disabled interrupt micros() could probably solve the
issue did you have a test?

**Comment by martinbudden** [comment]
Interesting. Perhaps sonar should used `SysTick->VAL` directly and handle rollover itself?

**Comment by jflyper** [comment]
@bk79 
No, I just picked this up this morning.

@martinbudden 
May be a good idea. A supplemental miros() call may be necessary to avoid two rollovers?

EDIT: Have to make usTicks in system.c global, or, create something like

```
uint32_t cycleToMicros(uint32_t cycles)
{
    return cycles/usTicks;
}
```

in system.c to keep usTicks static.

**Comment by martinbudden** [comment]
@jflyper sonar range is max 4m. So the max time for a sonar return pulse is 2\* 29.4 \* 400 = 23520 microseconds (sound travels at about 29.4 cm/us). So we should only ever get 1 rollover.

**Comment by jflyper** [comment]
@martinbudden 
What if a falling edge of an echo signal is missed for some 'really bad' reason?

**Comment by bk79** [comment]
3.5m max...over that tested on arduino you get rubbish and if you don't
have a smooth surface or you have a soft sound absorbing one...garbage as
well...so probably it's better to limit it at 2 m

**Comment by martinbudden** [comment]
Yes, limiting to less than 4m makes sense. Because the valid range for a pulse time is small, it should be fairly straightforward to detect missed echoes and ignore them.

**Comment by jflyper** [comment]
Okay, so it looks like the micros() called from ISR may be 1msec (little less) older than the actual timestamp. It explains occasional 30+cm spikes in the sonar measurement; 1msec older timestamp at rising edge, true timestamp at falling edge = total 1msec longer pulse.

Still, I recall I have seen 300cm+ measurement at 10cm actual. I wonder if I mis-read the 30+ as 300...

**Comment by digitalentity** [comment]
You didn't misread anything, sonar is very susceptible to vibration and noise. I've been playing with sonars for quite a while and I was unable to find one that will behave on a miniquad - all of them seem to read garbage after certain throttle value.

**Comment by stronnag** [comment]
I was offered (for a significant sum) a LightWare SF11/C today (laser rangefinder to 120m). Interesting device, but currently at a cost of more than I've ever spent building a 'copter. Unlikely to be tempted.

**Comment by jflyper** [comment]
Ah, I can understand that; vibe are velocities.

Time for optical?

**Comment by bk79** [comment]
260$ for halfing the baro precision...mmm...probably not worth

**Comment by jflyper** [comment]
LIDARs are overkill.

There's an IR-LED based 0-15m solution at around $115 
[LeddarOne](http://www.robotshop.com/en/leddartech-leddarone-optical-rangefinder-33v-uart.html), but it still is overkill.

I think I'm going to try combine several cheap IR-LED based devices;
- 100cm-500cm GP2Y0A710K0F
  http://www.sharp.co.jp/products/device/doc/opto/gp2y0a710k_e.pdf
  (There's ambiguity below 100cm, which can be cancelled by combining with something like below.)
- ~200cm
  [VL53L0X Time-of-Flight Distance Sensor Carrier with Voltage Regulator, 200cm Max](https://www.pololu.com/product/2490)

**Comment by martinbudden** [comment]
There's also the Teraranger http://www.robotshop.com/en/teraranger-one-tof-rangefinder-type-a-spider.html

**Comment by jflyper** [comment]
@martinbudden Look nice. Got into my list of optical range finders.

**Comment by digitalentity** [comment]
Optical is nice, however I still don't have the time to improve the sonar logic in INAV. Probably after #543

**Comment by jflyper** [comment]
#677 should improve things a little bit, if not at all.

I still want to evaluate vib effects digitalentity mentioned, though.

Ah, and some optical stuff, may be.

Closing for now.

**Comment by digitalentity** [comment]
@jflyper yes, #677 should fix issues with timing, but poor support of sonar altitude is deeper than that. At the momemt my time has beed diverted to better supporting fixed wing aircraft.
